test/filesystem.cc

```diff
@@ -135,6 +137,70 @@ TEST(Filesystem, extension) {
 TEST(Filesystem, file_size) {
 }
 
+TEST(Filesystem, concurrent_folder_create_delete) {
+
+  const char* create_subdir = "test/a/b/c/d/e/f/g/h/i/j";
+  const char* base_subdir = "test/a";
+
+  auto create_folder = [&]() {
+    bool success = filesystem::create_directories(create_subdir);
+    EXPECT_TRUE(success);
+    EXPECT_TRUE(filesystem::exists(create_subdir));
+  };
+
+  auto remove_folder = [&]() {
+    std::uintmax_t delete_count = filesystem::remove_all(base_subdir);
+    // I've found that two+ threads will claim that they've deleted the same
+    // file object. This results in double+ counting and prevents me from
+    // asserting anything related to the delete_count. All we can be sure
+    // is the following:
+    EXPECT_TRUE(!filesystem::exists(base_subdir));
+  };
+
+  size_t num_threads = std::thread::hardware_concurrency();
```

Comment:
Curious what other people think about `hardware_concurrency` in a test?

Since we're not compute bound here, there's not really an use in using the native number of threads, and it likely leads to the test behaving differently when running on a 2 core intel vs a 32 core amd.

Reply:
its not good. if CI doesnt have the resources to do it will pretty much fork bomb the machine.

Reply:
perhaps we can give it the disabled label so that the test exists but we dont run it in CI

Reply:
I was thinking of a fixed number of threads so it's independent of the machine it runs on.

Is resources an issue for this test? It's not forking the process, simply running the filesystem operation in a lightweight thread in this test, right?

Reply:
Here are my thoughts.  These are just threads - so quite lightweight.  Each thread is doing a trivial number of trivial things on disk.  The only memory consumption of note is the weight of each thread.  Creating 100 threads is only a touch slower than 8.  (I timed it.)

Anders' suggestion to "start the race" is a good one.  I mean, there are a ton of variables outside our control to force/guarantee concurrency - but we can control that one.  I'll implement that.

I don't have strong feelings either way regarding using a fixed number of threads vs the number available in the system.

Reply:
yeah i didnt literally mean fork bomb but more so the idea. i would agree if we can rewrite it some how to force trhe race that would be better.

there is a  huge problem with creating large numbers of threads which is that ive seen it cause problems on CI. even with larger machines on CI we use full parallelism when we run `make check` this means that all the threads that we could be using, we are already using. if you then start this test you are doubling that, which causes cpu contention and can lead to CI stalling out. it has happend to us many times in the past. since these are short lived threads maybe it wont be a problem but random CI failures isnt fun for anybody.

Reply:
Fair points K1.  Maybe we should run "make check" with less than all available cores/threads?  The runtime of make-check might increase a bit, but having the extra processing headroom should theoretically help cut down on stalls.

Honestly, in my experience, hyperthreads do help with scaling but not as much as you'd wish/want.  I might even go so far as to suggest running make-check with half the number of available system "cores", which is (at least in this day/age) the actual number of real cpu cores.

Combining K1's and my thoughts, I'm thinking that this test should only use half of the available system cores.

Curious everyone's thoughts on configuring CI's make-check to use half the available system cores?  Or other thoughts?

Reply:
> Curious everyone's thoughts on configuring CI's make-check to use half the available system cores? Or other thoughts?

Hmm maybe? I don't have enough to form an opinion here. As long as we don't decrease the number of processes used for compiling code.

Reply:
I've reduced the number of worker threads to num-available / 2.  On a separate channel we can continue thoughts on how many cores/threads to use when CI is running tests.